modal/_container_entrypoint.py

```diff
@@ -176,49 +176,53 @@ def get_max_inputs_to_fetch(self):
 
         return math.ceil(RTT_S / max(self.get_average_call_time(), 1e-6))
 
-    async def _generate_inputs(
-        self,
-    ) -> AsyncIterator[tuple[str, api_pb2.FunctionInput]]:
+    async def _generate_inputs(self) -> AsyncIterator[tuple[str, api_pb2.FunctionInput]]:
         request = api_pb2.FunctionGetInputsRequest(function_id=self.function_id)
         eof_received = False
         iteration = 0
         while not eof_received:
             request.average_call_time = self.get_average_call_time()
             request.max_values = self.get_max_inputs_to_fetch()  # Deprecated; remove.
+            request.input_concurrency = self.semaphore.maxsize
 
+            # If number of active inputs is at max queue size, this will block.
+            await self._acquire_semaphore()
             with trace("get_inputs"):
                 set_span_tag("iteration", str(iteration))  # force this to be a tag string
                 iteration += 1
                 response = await retry_transient_errors(self.client.stub.FunctionGetInputs, request)
 
+            yielded = False
+
             if response.rate_limit_sleep_duration:
                 logger.info(
                     "Task exceeded rate limit, sleeping for %.2fs before trying again."
                     % response.rate_limit_sleep_duration
                 )
                 await asyncio.sleep(response.rate_limit_sleep_duration)
-                continue
-
-            if not response.inputs:
-                continue
-
-            for item in response.inputs:
-                if item.kill_switch:
-                    logger.debug(f"Task {self.task_id} input received kill signal.")
-                    eof_received = True
-                    break
-
-                # If we got a pointer to a blob, download it from S3.
-                if item.input.WhichOneof("args_oneof") == "args_blob_id":
-                    input_pb = await self.populate_input_blobs(item.input)
-                else:
-                    input_pb = item.input
-
-                yield (item.input_id, input_pb)
-
-                if item.input.final_input:
-                    eof_received = True
-                    break
+            elif response.inputs:
+                for item in response.inputs:
+                    if item.kill_switch:
+                        logger.debug(f"Task {self.task_id} input received kill signal.")
+                        eof_received = True
+                        break
+
+                    # If we got a pointer to a blob, download it from S3.
+                    if item.input.WhichOneof("args_oneof") == "args_blob_id":
+                        input_pb = await self.populate_input_blobs(item.input)
+                    else:
+                        input_pb = item.input
+
+                    # If yielded, allow semaphore to be released via enqueue_outputs
```

Comment:
why have two different paths to release the semaphore? seems complex vs having a single place that acquires and releases it (ideally using a context manager)

Reply:
The path separate occurs because we need to yield the input - and the input is run either in a separate task or thread. Then, when the input finishes, we enqueue the output - and that happens from that separate task (async case).

(same reason as why we can't use a context manager)

Reply:
why do you have to acquire and release the semaphore in those places? isn't it fine to just do it surrounding calling the user function? i.e. `res = fun(*args, **kwargs)`

Reply:
The challenge is that the semaphore is acquired prior to the fetch, which happens in the generator. This will be included in the refactor proposal.